### PR TITLE
feat: adoption probes

### PR DESCRIPTION
**Pull request by alexey-yarmosh**
Fixes https://github.com/jsdelivr/globalping/issues/425
This functionality requires a connection to MariaDB with the 'adopted_probes' table configured.

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -8,3 +8,14 @@ services:
     volumes:
       - ./config/redis.conf:/usr/local/etc/redis/redis.conf
     command: /usr/local/etc/redis/redis.conf
+
+  mariadb:
+    image: mariadb:latest
```
Comment (MartinKolarik) [inline docker-compose.yml:13]: We should match the prod version here.

File: src/adoption-code/schema.ts [hunk 2]
```diff
@@ -0,0 +1,6 @@
+import Joi from 'joi';
+
+export const schema = Joi.object({
+	ip: Joi.string().ip().required(),
```
Comment (MartinKolarik) [inline src/adoption-code/schema.ts:4]: Should use `globalIpOptions`

File: config/init.sql [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+  userId VARCHAR(255) NOT NULL,
+  ip VARCHAR(255) NOT NULL,
+  uuid VARCHAR(255),
+  lastSyncDate DATE
+);
+INSERT IGNORE INTO adopted_probes (id, userId, ip) VALUES ('1', '6191378', '79.205.97.254');
+
+CREATE USER IF NOT EXISTS 'directus';
+GRANT ALL PRIVILEGES ON `directus`.* TO 'directus';
+FLUSH PRIVILEGES;
```
Comment (MartinKolarik) [inline config/init.sql:19]: This is dev only and in production, we rely on directus migrations, right?

File: src/lib/adopted-probes.ts [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+	}
+
+	private async syncDashboardData () {
+		const allSockets = await this.fetchWsSockets();
+		this.connectedIpToUuid = new Map(allSockets.map(socket => [ socket.data.probe.ipAddress, socket.data.probe.uuid ]));
+		this.connectedUuidToIp = new Map(allSockets.map(socket => [ socket.data.probe.uuid, socket.data.probe.ipAddress ]));
+
+		const adoptedProbes = await this.sql(TABLE_NAME).select<AdoptedProbe[]>('ip', 'uuid', 'lastSyncDate');
+		await Promise.all(adoptedProbes.map(({ ip, uuid }) => this.syncProbeIds(ip, uuid)));
+		await Promise.all(adoptedProbes.map(({ ip, lastSyncDate }) => this.updateSyncDate(ip, lastSyncDate)));
```
Comment (MartinKolarik) [inline src/lib/adopted-probes.ts:40]: Use `Bluebird.map()` with limited concurrency.

File: test/tests/unit/adopted-probes.test.ts [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+		fetchSocketsStub.resolves([{ data: { probe: { ipAddress: '1.1.1.1', uuid: '1-1-1-1-1' } } }]);
+
+		adoptedProbes.scheduleSync();
+		await sandbox.clock.tickAsync(5000);
+
+		expect(whereStub.callCount).to.equal(0);
+		expect(updateStub.callCount).to.equal(0);
+		expect(deleteStub.callCount).to.equal(0);
+	});
+});
```
Comment (MartinKolarik) [inline test/tests/unit/adopted-probes.test.ts:155]: No point in rewriting it now but note that here I'd greatly prefer integration tests that prepare data in a real DB, run the sync, and check the results. Would be much cleaner than all the mocking and provide better coverage as well.

File: config/init.sql (hunk 3 above)
Comment (alexey-yarmosh) [reply config/init.sql:19]: yes, directus uses `directus` user so it will be in place in production. maybe we should also have a specific user for gp api?

Comment (MartinKolarik) [reply config/init.sql:19]: User can be the same.

File: src/lib/adopted-probes.ts [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+		private readonly sql: Knex,
+		private readonly fetchWsSockets: typeof fetchSockets,
+	) {}
+
+	scheduleSync () {
+		setTimeout(() => {
+			this.syncDashboardData()
+				.finally(() => this.scheduleSync())
+				.catch(error => logger.error(error));
+		}, 5000);
```
Comment (MartinKolarik) [inline src/lib/adopted-probes.ts:30]: If we keep this running in multiple processes, let's bump the sync interval to 1 minute.

**Comment by MartinKolarik** [comment]
Even though we don't directly test it, the process is trying to connect to the DB during tests. This previously didn't fail only because the timeout was 1 minute and the tests run faster then that:

image

Now, it fails after I reduced the connection timeout. We need to either prevent this or configure the DB on CI (https://github.com/jsdelivr/data.jsdelivr.com/blob/master/.github/workflows/nodejs.yml#L27-L32).

**Pull request merged**